## Goal

Hey SolveIt, I am Tom Liu, a python developer, and I want to practice coding skills and cracking Advent of Code problems.

## Plan
* Devide and conque with breaking the problems to small parts, then exercise problem solving on small ones
* Ask right questions when solving each problems
* Reflect and iterate the problem solving.

## --- Day 2: Red-Nosed Reports ---

Title: Day 2: Red-Nosed Reports

Fortunately, the first location The Historians want to search isn't a long walk from the Chief Historian's office.

While the Red-Nosed Reindeer nuclear fusion/fission plant appears to contain no sign of the Chief Historian, the engineers there run up to you as soon as they see you. Apparently, they still talk about the time Rudolph was saved through molecular synthesis from a single electron.

They're quick to add that - since you're already here - they'd really appreciate your help analyzing some unusual data from the Red-Nosed reactor. You turn to check if The Historians are waiting for you, but they seem to have already divided into groups that are currently searching every corner of the facility. You offer to help with the unusual data.

The unusual data (your puzzle input) consists of many reports, one report per line. Each report is a list of numbers called levels that are separated by spaces. For example:

7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
This example data contains six reports each containing five levels.

The engineers are trying to figure out which reports are safe. The Red-Nosed reactor safety systems can only tolerate levels that are either gradually increasing or gradually decreasing. So, a report only counts as safe if both of the following are true:

The levels are either all increasing or all decreasing.
Any two adjacent levels differ by at least one and at most three.
In the example above, the reports can be found safe or unsafe by checking those rules:

7 6 4 2 1: Safe because the levels are all decreasing by 1 or 2.
1 2 7 8 9: Unsafe because 2 7 is an increase of 5.
9 7 6 2 1: Unsafe because 6 2 is a decrease of 4.
1 3 2 4 5: Unsafe because 1 3 is increasing but 3 2 is decreasing.
8 6 4 4 1: Unsafe because 4 4 is neither an increase or a decrease.
1 3 6 7 9: Safe because the levels are all increasing by 1, 2, or 3.
So, in this example, 2 reports are safe.

Analyze the unusual data from the engineers. How many reports are safe?



In [ ]:
from aocd.models import Puzzle
p = Puzzle(year=2024, day=2)
p.examples

[Example(input_data='7 6 4 2 1\n1 2 7 8 9\n9 7 6 2 1\n1 3 2 4 5\n8 6 4 4 1\n1 3 6 7 9', answer_a='2', answer_b=None, extra=None)]

In [ ]:
inp = p.examples[0].input_data
inp

'7 6 4 2 1\n1 2 7 8 9\n9 7 6 2 1\n1 3 2 4 5\n8 6 4 4 1\n1 3 6 7 9'

convert the string to be a list with 5 int list

In [ ]:
report = [line.split().map(int) for line in inp.splitlines()]
report

AttributeError: 'list' object has no attribute 'map'

how to convert a string array to be int array

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

In Python, `list` doesn't have a `.map()` method like in some other languages. Here are two common ways to convert a list of strings to integers:

**Option 1: List comprehension** (most Pythonic)
```python
[int(x) for x in line.split()]
```

**Option 2: Using `map()` with `list()`**
```python
list(map(int, line.split()))
```

Both do the same thing! Which style do you prefer to use in your code?

let's do option 2

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great! Here's how to fix your code with option 2:

```python
report = [list(map(int, line.split())) for line in inp.splitlines()]
report
```

Try running this and see what you get!

In [ ]:
reports = [list(map(int, line.split())) for line in inp.splitlines()]
reports

[[7, 6, 4, 2, 1],
 [1, 2, 7, 8, 9],
 [9, 7, 6, 2, 1],
 [1, 3, 2, 4, 5],
 [8, 6, 4, 4, 1],
 [1, 3, 6, 7, 9]]

In [ ]:
row = reports[0]
row

[7, 6, 4, 2, 1]

In [ ]:
diff_a = []
for i in range(0, len(row) - 1):
    diff_a.append(row[i + 1] - row[i])
diff_a

[-1, -2, -2, -1]

In [ ]:
(all([item > 0  for item in diff_a]) or all([item < 0  for item in diff_a])) and all([abs(item) <= 3 for item in diff_a])

True

In [ ]:
# define the safe report checking
def safe_report(report):
    diff_a = []
    for i in range(0, len(report) - 1):
        diff_a.append(report[i + 1] - report[i])
    if (all([item > 0  for item in diff_a]) or all([item < 0  for item in diff_a])) and all([abs(item) <= 3 for item in diff_a]):
        return 1
    else:
        return 0

In [ ]:
safe_report(row)

True

In [ ]:
for report in reports:
    print(safe_report(report))

1
0
0
0
0
1


In [ ]:
# get the reports
def get_reports(input):
    reports = [list(map(int, line.split())) for line in input.splitlines()]
    return reports


In [ ]:
all_reports = get_reports(p.input_data)
all_reports[1]

[3, 5, 7, 10, 11, 11]

hey solveit, I will process safe/unsafe checking on all reports, and will get the result on each report: 1 means safe, and 0 means unsafe, then I want to sum the result to get the number of safe reports. 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Nice plan! You can use a list comprehension to apply `safe_report()` to each report, then use `sum()` to count the safe ones:

```python
sum([safe_report(report) for report in all_reports])
```

This works because `sum()` adds up all the 1s and 0s. Give it a try!

In [ ]:
sum([safe_report(report) for report in all_reports])

421

awesome. that's the correct answer. Now, please coach me to use numpy to solve the problem.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent work! Let's solve this using NumPy. The key advantage is that NumPy can compute differences between adjacent elements very efficiently.

First, let's look at how to calculate the differences for a single report using NumPy:

```python
import numpy as np
diffs = np.diff(row)
diffs
```

Try this out and see what you get. How does it compare to your `diff_a` list from before?

In [ ]:
import numpy as np
diffs = np.diff(row)
row, diffs

([7, 6, 4, 2, 1], array([-1, -2, -2, -1]))

awesome. that's exactly the same result I did with for loop. given the rules of safe report, I want to check all negative, or all positive, and especially, the absolute value of each diff is less or equal to 3. any suggestions?


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great! Now you can use NumPy's boolean operations to check those conditions efficiently:

For checking all positive or all negative:
```python
(diffs > 0).all() or (diffs < 0).all()
```

For checking absolute values ≤ 3:
```python
(np.abs(diffs) <= 3).all()
```

Try these out separately on `diffs` to see what they return. How would you combine them to check if a report is safe?

In [ ]:
(diffs > 0).all() or (diffs < 0).all()

np.True_

In [ ]:
(np.abs(diffs) <= 3).all()

np.True_

In [ ]:
def safe_report2(report):
    diffs = np.diff(report)
    return ((diffs > 0).all() or (diffs < 0).all()) and (np.abs(diffs) <=3).all()

safe_report2(row)

np.True_

Hey solveit, that's how I would like to construct safe report checking using numpy. please review, and provide your suggestions.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Your `safe_report2()` function looks great! It's clean and uses NumPy efficiently. 

One small thing to note: there's a subtle issue with `(np.abs(diffs) <=3).all()` - you need to add `()` around `diffs` when checking the absolute value condition:

```python
(np.abs(diffs) <= 3).all()
```

Actually, looking at your code again, you already have it correct! The function should work perfectly.

Now, do you want to test it on all the reports to see if you get the same answer (421)?

let's do it.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! Try this:

```python
sum([safe_report2(report) for report in all_reports])
```

Run it and see if you get 421 again!

In [ ]:
sum([safe_report2(report) for report in all_reports])

np.int64(421)

Brilliant! I got the same answer. Wow, numpy is much efficient. Now, please review my problem solving process. As you know, I want to practice solveit method with breaking a big problem to be smaller ones, and tackle them one by one. any feedback or coaching suggestions?


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent work, Tom! Your problem-solving approach was really solid. Here's what you did well:

**Strengths:**
- You broke down the problem into clear steps: parse input → understand one report → check safety rules → apply to all reports
- You tested on a single example (`row`) before scaling up to all reports
- You iterated and improved your solution (loop-based → NumPy-based)

**Suggestions for next time:**
- Consider writing out the sub-problems explicitly before coding (like a checklist)
- For each sub-problem, you could create a small test case to verify it works before moving on

One question: When you were working through this, did you consciously identify the sub-problems first, or did they emerge as you coded? Understanding your natural approach can help you refine it further.

very good suggestion and question. I am trying to copying Jeremy and Johno practice to identify sub-problems. 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

That's a great approach to model! Breaking problems into sub-problems is a skill that gets better with practice.

One technique that can help: before writing any code, try writing out the sub-problems as comments or notes. For example, for this problem you might have written:

```
1. Parse input into list of reports
2. Calculate differences between adjacent levels
3. Check if all increasing OR all decreasing
4. Check if all differences are 1-3
5. Count safe reports
```

Would you like to try this on the next Advent of Code problem? Or do you have questions about how to identify sub-problems more systematically?

Yes, indeed. I want to try it on the next one. One more, any suggestions on identifying sub-problems more systematically? I believe that I should do "practices make perfect"!

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great mindset! Here are a few systematic approaches:

**1. Work backwards:** Start with the final answer you need, then ask "what do I need to have ready to calculate this?"

**2. Look for transformations:** Input → intermediate form → another form → final answer. Each transformation is a sub-problem.

**3. Test-driven thinking:** What's the smallest thing you could test? That's often your first sub-problem.

For Advent of Code specifically, the examples are gold - they often reveal the sub-problems naturally.

Ready to tackle Day 2 Part 2, or do you want to try a different day?

Let's continue to tackle Day 2 Part 2, and the details:

## --- Part Two ---
The engineers are surprised by the low number of safe reports until they realize they forgot to tell you about the Problem Dampener.

The Problem Dampener is a reactor-mounted module that lets the reactor safety systems tolerate a single bad level in what would otherwise be a safe report. It's like the bad level never happened!

Now, the same rules apply as before, except if removing a single level from an unsafe report would make it safe, the report instead counts as safe.

More of the above example's reports are now safe:

7 6 4 2 1: Safe without removing any level.
1 2 7 8 9: Unsafe regardless of which level is removed.
9 7 6 2 1: Unsafe regardless of which level is removed.
1 3 2 4 5: Safe by removing the second level, 3.
8 6 4 4 1: Safe by removing the third level, 4.
1 3 6 7 9: Safe without removing any level.
Thanks to the Problem Dampener, 4 reports are actually safe!

Update your analysis by handling situations where the Problem Dampener can remove a single level from unsafe reports. How many reports are now safe?

In [ ]:
inp = p.examples[0]
inp

Example(input_data='7 6 4 2 1\n1 2 7 8 9\n9 7 6 2 1\n1 3 2 4 5\n8 6 4 4 1\n1 3 6 7 9', answer_a='2', answer_b='4', extra=None)

In [ ]:
reports = np.array([np.array(line.split(), np.int64) for line in inp.input_data.splitlines()])
reports

array([[7, 6, 4, 2, 1],
       [1, 2, 7, 8, 9],
       [9, 7, 6, 2, 1],
       [1, 3, 2, 4, 5],
       [8, 6, 4, 4, 1],
       [1, 3, 6, 7, 9]])

Hey solveit, I understand that the correct answer for `$reports` is 4, but how can I design test-driven to implement my solution. what should I start with?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great question! Let's think test-driven. Before writing the full solution, what's the smallest thing you could verify?

I'd suggest starting with a single report and testing if it's safe with the Problem Dampener. For example:

**Test case:** `[1, 3, 2, 4, 5]`
- Currently unsafe (direction changes)
- Should be safe if you remove the `3` (index 1)
- Expected result: safe (True)

Want to start by writing a function that tries removing each level one at a time and checks if any resulting report is safe? What would you call this function?Great question! Let's start with the smallest testable piece.

Before writing the full solution, what's the first thing you need to be able to do? 

Here's a hint: if a report is unsafe, you need to try removing each level one at a time. So maybe start with:

**Sub-problem 1:** Given a report and an index, can you create a new report with that index removed?

Try writing a small function or one-liner that takes `reports[3]` (which is `[1, 3, 2, 4, 5]`) and removes the element at index 1 (the `3`). What do you get?

In [ ]:
def safe_with_dampener(report):
    # First check: is it already safe?
    if safe_report2(report):
        return True
    # TODO: handle the dampener logic
    return False

safe_with_dampener(reports[0])  # Should return True

True

In [ ]:
safe_with_dampener(reports[3])

False

In [ ]:
diffs = np.diff(reports[3])
diffs

array([ 2, -1,  2,  1])

In [ ]:
# to check that in all increasing levels, whether report may be the one to remove one level and test it again.
np.sum(diffs > 0) == (diffs.shape[0] - 1)

np.True_

now, I want to break it down to sub-problems, when we have Dampener:
- we need to identify reports that can with one 'tolerated a single bad level'
- when reports fit to part 1 requirement, they are safe.
- when reports don't fit to part 1's, we check the `diffs`
  - if there 3 elements under the diffs fit to the part 1's , we need to identify outliner and remove it, then test on part 1's again.
  - otherwise, the reports is unsafe

Is these problem breakdown good?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good start on breaking it down! I see your thinking, but there's a simpler approach that might be more reliable.

Your current approach tries to be clever about *which* level to remove. But here's a question: how many levels are in each report? Usually 5-8, right?

What if instead of trying to identify the outlier, you just try removing *each* level one at a time and test if any resulting report is safe? 

This is simpler because:
- You already have `safe_report2()` working
- No need to analyze diffs to find the outlier
- Brute force, but fast enough for small lists

What do you think - want to try that approach first?

Wow, that's a good thought! Let me work on 'brute force' method first.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! Let's do it.

Start with the sub-problem we discussed: given a report and an index, remove that element. 

Try this for `reports[3]` (which is `[1, 3, 2, 4, 5]`) - remove the element at index 1:

```python
np.delete(reports[3], 1)
```

What do you get? Does it look like what you'd expect?

In [ ]:
test_report = np.delete(reports[3], 1)
test_report 

array([1, 2, 4, 5])

In [ ]:
safe_report2(test_report)

np.True_

I want to avoid using for loop to delete an element one by one, how about duplicate the report with a matrix with row equals to the amount of elements, then delete the diagonal element from the matrix, and run the `safe_report2` checking on each row? thoughts?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

That's a creative vectorized approach! It would definitely avoid explicit loops.

However, there's a challenge: `safe_report2` expects a 1D array, but after deleting diagonal elements, you'd have rows of different "shapes" (conceptually). NumPy matrices need uniform dimensions.

You could work around this, but honestly - for Advent of Code with reports of ~5-8 elements, a simple loop would be clearer and fast enough. Sometimes simple is better!

Want to try the loop approach first and see how it performs? Or do you want to explore the matrix idea further?

In [ ]:
# let's do for loop for simplicity
def with_problem_damper_check(report):
    for i in range(0, report.shape[0]):
        tweaked_report = np.delete(report, i)
        safe = safe_report2(tweaked_report)
        if safe: # return True as we found the 'tolerated single bad level'
            return np.True_
    return np.False_

def safe_with_dampener(report):
    # First check: is it already safe?
    if safe_report2(report):
        return True
    
    # handle the dampener logic
    return with_problem_damper_check(report)

safe_with_dampener(reports[0]), safe_with_dampener(reports[1]), safe_with_dampener(reports[3])  # Should return True, False, True




(True, np.False_, np.True_)

hey solveit, could you please review `with_problem_damper_check` per our discussion?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Your `with_problem_damper_check` function looks solid! It does exactly what we discussed - tries removing each element and checks if any result is safe.

A couple of small notes:

1. You can simplify the return - just use `return safe` instead of the if/else, since `safe` is already a boolean
2. The final `return np.False_` is correct - it means no single removal made it safe

Want to test it on all the example reports to see if you get the expected answer of 4?

In [ ]:
# let's do for loop for simplicity
def with_problem_damper_check(report):
    for i in range(0, report.shape[0]):
        tweaked_report = np.delete(report, i)
        safe = safe_report2(tweaked_report)
        if safe: # return True as we found the 'tolerated single bad level'
            return np.True_
    return safe_report2(report)

with_problem_damper_check(reports[0]), with_problem_damper_check(reports[1]), with_problem_damper_check(reports[3])  # Should return True, False, True




(np.True_, np.False_, np.True_)

I fixed it with your feedback. please review.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good effort, but there's a logical issue now!

Your function returns `safe_report2(report)` at the end - but that checks the *original* report. If the loop didn't find a safe version by removing one element, the report should be unsafe (return `False`).

The original version with `return np.False_` was actually correct! Does that make sense?

In [ ]:
# let's do for loop for simplicity
def with_problem_damper_check(report):
    if safe_report2(report):
        return np.True_
    for i in range(0, report.shape[0]):
        tweaked_report = np.delete(report, i)
        safe = safe_report2(tweaked_report)
        if safe: # return True as we found the 'tolerated single bad level'
            return np.True_
    return np.False_

with_problem_damper_check(reports[0]), with_problem_damper_check(reports[1]), with_problem_damper_check(reports[3])  # Should return True, False, True




(np.True_, np.False_, np.True_)

please review again.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Nice! Your function now works correctly. The logic is clear:

1. Check if already safe → return True
2. Try removing each element → if any is safe, return True  
3. If nothing worked → return False

Now you're ready to test it on all the example reports. Want to run it and see if you get 4 safe reports?

In [ ]:
sum([safe_with_dampener(report) for report in reports])

np.int64(4)

I've verified it on example and it's correct.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent! You got the expected answer of 4 on the example data. 

Now you're ready to run it on the full puzzle input. Want to try it on `all_reports` (the full data from `p.input_data`)?

In [ ]:
all_reports = np.array([np.array(line.split(), np.int64) for line in p.input_data.splitlines()])

all_reports

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (1000,) + inhomogeneous part.

what does the error mean?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

The error means the reports in the full data have different lengths (some might have 5 numbers, others 6, 7, etc.), so NumPy can't create a uniform 2D array.

For the example data, all reports happened to be length 5, but the real data varies!

The simple fix: keep it as a Python list instead of a NumPy array:

```python
all_reports = [np.array(line.split(), np.int64) for line in p.input_data.splitlines()]
```

This way each report is still a NumPy array (so `safe_with_dampener` works), but they're in a list. Try it!

In [ ]:
all_reports = [np.array(line.split(), np.int64) for line in p.input_data.splitlines()]

all_reports

[array([90, 91, 93, 96, 93]),
 array([ 3,  5,  7, 10, 11, 11]),
 array([35, 37, 39, 42, 46]),
 array([67, 70, 72, 74, 79]),
 array([ 9, 12, 13, 16, 15, 16, 19]),
 array([48, 51, 52, 55, 58, 61, 58, 57]),
 array([3, 4, 7, 9, 8, 9, 9]),
 array([22, 25, 28, 30, 28, 32]),
 array([38, 41, 44, 45, 42, 49]),
 array([54, 57, 59, 59, 61]),
 array([54, 55, 55, 56, 58, 55]),
 array([66, 68, 70, 71, 72, 72, 75, 75]),
 array([32, 35, 35, 38, 42]),
 array([22, 24, 26, 28, 28, 31, 34, 41]),
 array([55, 57, 59, 63, 64]),
 array([18, 21, 25, 26, 28, 27]),
 array([39, 40, 41, 43, 47, 49, 50, 50]),
 array([10, 11, 14, 17, 20, 22, 26, 30]),
 array([35, 36, 39, 40, 44, 45, 52]),
 array([57, 59, 66, 67, 70]),
 array([15, 17, 19, 22, 29, 26]),
 array([66, 69, 72, 75, 78, 83, 83]),
 array([73, 76, 78, 83, 86, 90]),
 array([23, 26, 28, 35, 36, 41]),
 array([83, 81, 82, 85, 88, 90, 92, 93]),
 array([4, 2, 3, 4, 1]),
 array([ 8,  7,  8,  9, 11, 13, 13]),
 array([71, 70, 71, 73, 76, 80]),
 array([22, 20, 21, 22, 

In [ ]:
sum([safe_with_dampener(report) for report in all_reports])

np.int64(476)

awesome. I've got the correct answer, with numpy! could you please review my part 2 problem solving process, and provide feedback/suggestions?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great work, Tom! You successfully solved Part 2 using NumPy. Here's my review:

**What you did well:**
- You started with a clear function structure (`safe_with_dampener`) and tested it incrementally
- You reused your existing `safe_report2` function effectively
- You implemented the brute-force approach cleanly - trying each removal systematically
- You tested on example data before running on the full dataset

**Areas for improvement:**
- **Sub-problem identification**: You jumped into coding fairly quickly. Did you explicitly write out the sub-problems before coding, or did they emerge as you went?
- **Edge case thinking**: Before coding, did you consider: "What if a report is already safe?" (You handled this, but was it planned upfront?)

**Question for reflection:**
Looking back, if you were to tackle a similar problem tomorrow, what would you do differently in your planning phase?

good feedback! I was stuck on sub-problem identification - should I use brute-force method (which I don't like as it potentially may cost time), or evaluate the report whether it fits to tolerate a single bad level. Then, I asked solveit on guidance; for 'edge case thinking', I thought about it, and iterate the method to check it first using `safe_report2`. 

overall, you are right that I think too quickly and should practice sub-problem breakdown.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

That's excellent self-awareness, Tom! You identified a key tension: **efficiency vs. simplicity**.

Here's a useful heuristic for Advent of Code (and coding in general):

**Start simple, optimize if needed.** 

For small inputs (like ~1000 reports with ~8 levels each), brute force is totally fine. If it were 1 million reports, then you'd optimize.

**Practice idea:** Next time you face this tension, try writing down:
1. The simple approach (brute force)
2. The optimized approach (if you can think of one)
3. Estimate: will the simple approach be fast enough?

This helps you make an informed decision rather than getting stuck.

Ready to tackle another problem, or want to reflect more on this one?

hey solveit, now, please review my whole problem solving process for part 1 and 2, then show your final feedback.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great job on completing both parts, Tom! Here's my overall feedback:

**Strengths:**
- **Incremental testing**: You tested on single examples before scaling up - this is excellent practice
- **Code reuse**: You built on `safe_report` to create `safe_report2`, then reused it in Part 2
- **Iteration**: You refined your approach based on feedback (NumPy version, edge cases)
- **Tool selection**: You successfully applied NumPy to make your code more concise

**Areas to develop:**
- **Upfront planning**: Write out sub-problems as a checklist before coding - this helps you see the full picture
- **Decision framework**: When facing efficiency trade-offs, estimate costs before getting stuck
- **Test-driven start**: For Part 2, you could have written test cases for the examples first, then built the solution

**One challenge for next time:**
Before writing any code, spend 2-3 minutes writing comments outlining your approach. Try it and see if it changes how you code!

How do you feel about your progress? What's one thing you want to focus on for the next problem?